src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,243 @@
+//! Generalized HTTP scrape source.
+//! Scrapes an endpoint at an interval, decoding the HTTP responses into events.
+
+use bytes::{Bytes, BytesMut};
+use chrono::Utc;
+use futures_util::FutureExt;
+use http::{response::Parts, Uri};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use tokio_util::codec::Decoder as _;
+
+use crate::{
+    codecs::{Decoder, DecodingConfig},
+    config::{SourceConfig, SourceContext, SourceDescription},
+    http::Auth,
+    serde::default_decoding,
+    serde::default_framing_message_based,
+    sources,
+    tls::{TlsConfig, TlsSettings},
+    Result,
+};
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use vector_config::configurable_component;
+use vector_core::{
+    config::{log_schema, LogNamespace, Output},
+    event::Event,
+};
+
+/// The name of this source
+pub(crate) const NAME: &str = "http_scrape";
+
+/// Configuration for the `http_scrape` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+pub struct HttpScrapeConfig {
+    /// Endpoint to scrape events from. The full path must be specified.
+    /// Example: "http://127.0.0.1:9898/logs"
+    endpoint: String,
+
+    /// The interval between scrapes, in seconds.
+    #[serde(default = "super::default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+
+    /// Custom parameters for the scrape request query string.
+    ///
+    /// One or more values for the same parameter key can be provided. The parameters provided in this option are
+    /// appended to any parameters manually provided in the `endpoint` option.
+    query: Option<HashMap<String, Vec<String>>>,
+
+    /// Decoder to use on the HTTP responses.
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    decoding: DeserializerConfig,
+
+    /// Framing to use in the decoding.
+    #[configurable(derived)]
+    #[serde(default = "default_framing_message_based")]
+    framing: FramingConfig,
+
+    /// Headers to apply to the HTTP requests.
+    /// One or more values for the same header can be provided.
+    #[serde(default)]
+    headers: Option<HashMap<String, Vec<String>>>,
```

Comment:
I haven't read below yet. Do you make a distinction between no header map and an empty one? If not, the `Option` looks superfluous here. Note that the initial empty hash map allocates no memory.

Reply:
There is a consequence of doing that- unless there is something I'm not aware of, users will have to specify an empty map in their configuration files for query and header, which to me seems like a reason to keep the Option<..>.

Reply:
Scratch that, I just need the serde default on query.